### PR TITLE
Benchmark logging

### PR DESCRIPTION
**Pull request by II245**

### CODE REVIEW

File: benchmark.py [hunk 1]
```diff
@@ -40,16 +43,38 @@
...
-	args.model, args.lang, args.sample_rate, args.window_size, args.window_stride, args.window, args.num_input_features = map(checkpoint['args'].get, ['model', 'lang', 'sample_rate', 'window_size', 'window_stride', 'window', 'num_input_features'])
+	args.model, args.lang, args.sample_rate, args.window_size, args.window_stride, args.window, args.num_input_features = map(
+		checkpoint['args'].get,
+		['model', 'lang', 'sample_rate', 'window_size', 'window_stride', 'window', 'num_input_features'])
 
 use_cuda = 'cuda' in args.device
 
 labels = datasets.Labels(datasets.Language(args.lang))
 
+if args.profile_cuda:
```
Comment (vadimkantorov) [inline benchmark.py]: this would affect the profiled timings as well. please introduce a new argparse flag

File: scripts/profile.sh [hunk 2]
```diff
@@ -3,8 +3,9 @@ set -e
 TRACEFILE=data/profile.sqlite
 TRACELOG=data/profile.txt
 TRACEPYPROFLOG=data/profile.pyprof.txt
-
-nvprof -s -f -o $TRACEFILE --profile-from-start off -- python3 benchmark.py --profile-cuda $@ &> $TRACELOG
+LOGDEST=cudnn_log.txt
+CUBLAS_LOGDEST=cublas_log.txt
+CUDNN_LOGDEST_DBG=$LOGDEST CUBLAS_LOGDEST_DBG=$CUBLAS_LOGDEST nvprof -s -f -o $TRACEFILE --profile-from-start off -- python3 benchmark.py --profile-cuda $@ &> $TRACELOG
```
Comment (vadimkantorov) [inline scripts/profile.sh]: please set `LOGDEST*` in benchmark.py as well

Comment (vadimkantorov) [inline scripts/profile.sh]: also it's likely not very useful to profile with CUDNN/CUBLAS logs enabled, it can skew the timings, so don't do that

File: utils.py [hunk 3]
```diff
@@ -88,8 +89,12 @@ def compute_ram_memory_stats(byte_scaler =1024 ** 3):
 
 
 def compute_memory_fragmentation():
```
Comment (vadimkantorov) [inline utils.py:91]: please move this function to `utils.py`, it may get used elsewhere as well

File: utils.py [hunk 4]
```diff
@@ -88,8 +89,12 @@ def compute_ram_memory_stats(byte_scaler =1024 ** 3):
...
-	snapshot = torch.cuda.memory_snapshot()
```
Comment (vadimkantorov) [inline utils.py:91]: why did you remove the memory fragmentation code? it is still useful

File: utils.py [hunk 5]
```diff
@@ -88,8 +89,12 @@ def compute_ram_memory_stats(byte_scaler =1024 ** 3):
...
-	return sum(b['allocated_size'] for b in snapshot) / sum(b['total_size'] for b in snapshot)
+    lib = ctypes.cdll.LoadLibrary(None)
```
Comment (vadimkantorov) [inline utils.py]: if you want to compute device free/total memory stats, please introduce a new function, do not hijack an existing one

File: benchmark.py [hunk 6]
```diff
@@ -40,16 +43,38 @@
...
+			device_type = batch_ortvalue.device_name(),
+			device_id = 0,
+			element_type = batch.dtype,
+			shape = batch_ortvalue.shape(),
+			buffer_ptr = batch_ortvalue.data_ptr()
+		)
+		io_binding.bind_output(args.onnx_output_node_name, device)
+		return io_binding
+
+	load_batch = lambda batch: load_batch_to_device(batch, args.device)
```
Comment (vadimkantorov) [inline benchmark.py:82]: again, why do perform binding set up and CPU->GPU copy at every batch processing. let's do change benchmark.py to the same scheme we used in your other benchmark: move the batch to GPU before the loop and then just measure the GPU computation part.

File: utils.py [hunk 7]
```diff
@@ -88,8 +89,12 @@ def compute_ram_memory_stats(byte_scaler =1024 ** 3):
...
+    free_mem = ctypes.c_long()
+    total_mem = ctypes.c_long()
+    lib.cudaMemGetInfo(ctypes.byref(free_mem), ctypes.byref(total_mem))
+    return (total_mem.value - free_mem.value) / (1024 ** 3)
```
Comment (vadimkantorov) [inline utils.py]: either return bytes or move `1024 ** 3` to a `scaler` argument, like we have for `compute_model_capacity`

File: utils.py (hunk 4 above)
Comment (II245) [reply utils.py:91]: Because there is only one use of this function. Old function cause division by zero. The new one works for both cases onnx and pytorch.

Comment (vadimkantorov) [reply utils.py:91]: 1. please fix division by zero
2. the semantics of the old and new functions is completely different, so it does not make sense to change the implementation of the old one
3. feel free to add measurements with the new function as well

Comment (vadimkantorov) [reply utils.py:91]: the old function measures fragmentation, the new function measures utilisation. these are not the same thing

File: benchmark.py (hunk 1 above)
Comment (II245) [reply benchmark.py]: done

File: scripts/profile.sh (hunk 2 above)
Comment (II245) [reply scripts/profile.sh]: done

Comment (II245) [reply scripts/profile.sh]: done

File: utils.py (hunk 4 above)
Comment (II245) [reply utils.py:91]: function semantic is changed

File: utils.py (hunk 5 above)
Comment (II245) [reply utils.py]: answer abowe

File: utils.py (hunk 7 above)
Comment (II245) [reply utils.py]: function semantics is changed

File: benchmark.py (hunk 6 above)
Comment (II245) [reply benchmark.py:82]: fixed

File: utils.py [hunk 8]
```diff
@@ -91,6 +92,15 @@ def compute_memory_fragmentation():
 	snapshot = torch.cuda.memory_snapshot()
 	return sum(b['allocated_size'] for b in snapshot) / sum(b['total_size'] for b in snapshot)
 
+
+def compute_memory_fragmentation_ctypes():
```
Comment (vadimkantorov) [inline utils.py]: it does not measure memory fragmentation in the allocator. it measures memory utilisation. please rename the function to reflect its semantics, e.g. to `compute_global_device_memory_utilisation`

File: benchmark.py [hunk 9]
```diff
@@ -116,22 +148,26 @@
...
+	args.iterations)
+
+loaded_batch = load_batch(batch)
 for i in range(args.iterations):
 	tic = tictoc()
-	y = model(load_batch(batch))
+	y = model(loaded_batch)
 	toc = tictoc()
-	fragmentation[i] = utils.compute_memory_fragmentation()
+	fragmentation[i] = utils.compute_memory_fragmentation_ctypes() if args.onnx else utils.compute_memory_fragmentation()
```
Comment (vadimkantorov) [inline benchmark.py]: please introduce a new array for storing utilisation

File: benchmark.py [hunk 10]
```diff
@@ -116,22 +148,26 @@
...
+times_fwd, times_bwd, fragmentation = torch.zeros(args.iterations), torch.zeros(args.iterations), torch.zeros(
+	args.iterations)
+
+loaded_batch = load_batch(batch)
 for i in range(args.iterations):
 	tic = tictoc()
-	y = model(load_batch(batch))
+	y = model(loaded_batch)
 	toc = tictoc()
-	fragmentation[i] = utils.compute_memory_fragmentation()
```
Comment (vadimkantorov) [inline benchmark.py:134]: please keep the fragmentation measurement as well

File: benchmark.py [hunk 11]
```diff
@@ -116,22 +148,26 @@
...
+		use_cuda = use_cuda, enabled = True, profile_memory = True, record_shapes = True
+	)
 	autograd_profiler.__enter__()
 
 print('Starting benchmark for', args.iterations, 'iterations:', 'fwd', '+ bwd' if args.backward else '')
 tic_wall = tictoc()
-times_fwd, times_bwd, fragmentation = torch.zeros(args.iterations), torch.zeros(args.iterations), torch.zeros(args.iterations)
+times_fwd, times_bwd, fragmentation = torch.zeros(args.iterations), torch.zeros(args.iterations), torch.zeros(
+	args.iterations)
+
```
Comment (vadimkantorov) [inline benchmark.py:164]: the pin_memory call is now useless, please remove it

Comment (II245) [reply benchmark.py:164]: done

File: benchmark.py (hunk 10 above)
Comment (II245) [reply benchmark.py:134]: done

File: benchmark.py (hunk 9 above)
Comment (II245) [reply benchmark.py]: done

File: utils.py (hunk 8 above)
Comment (II245) [reply utils.py]: done
